rust/processor/src/db_writer.rs

```diff
@@ -0,0 +1,325 @@
+                                .inc();
+                            self.clean();
+                            res = self.execute_query(db_pool.clone()).await;
+                            // Go to next loop iteration to see if this was OK or Err
+                            continue;
+                        }
+                    }
+
+                    // Sleep a bit so we're not immediately coming back to this task
+                    // TODO: make this configurable?
+                    tokio::time::sleep(std::time::Duration::from_millis(50)).await;
+                    // Then we retry
+                    res = self.execute_query(db_pool.clone()).await;
+                },
+            }
+        }
+
+        DB_SINGLE_BATCH_EXECUTION_TIME_IN_SECS
+            .with_label_values(&[processor_name, table_name])
+            .set(insertion_timer.elapsed().as_secs_f64());
+
+        DB_INSERTION_ROWS_COUNT
+            .with_label_values(&[processor_name, table_name])
+            .inc_by(self.len() as u64);
+        res
+    }
+}
+
+pub fn diesel_error_to_metric_str(error: &Error) -> &'static str {
+    match error {
+        Error::DatabaseError(kind, _) => match kind {
+            DatabaseErrorKind::UniqueViolation => "DatabaseError::UniqueViolation",
+            DatabaseErrorKind::ForeignKeyViolation => "DatabaseError::ForeignKeyViolation",
+            DatabaseErrorKind::UnableToSendCommand => "DatabaseError::UnableToSendCommand",
+            DatabaseErrorKind::SerializationFailure => "DatabaseError::SerializationFailure",
+            DatabaseErrorKind::ReadOnlyTransaction => "DatabaseError::ReadOnlyTransaction",
+            DatabaseErrorKind::NotNullViolation => "DatabaseError::NotNullViolation",
+            DatabaseErrorKind::CheckViolation => "DatabaseError::CheckViolation",
+            DatabaseErrorKind::ClosedConnection => "DatabaseError::ClosedConnection",
+            DatabaseErrorKind::Unknown => "DatabaseError::Unknown",
+            _ => "DatabaseError::Other",
+        },
+        Error::NotFound => "NotFound",
+        Error::QueryBuilderError(_) => "QueryBuilderError",
+        Error::DeserializationError(_) => "DeserializationError",
+        Error::SerializationError(_) => "SerializationError",
+        Error::InvalidCString(_) => "InvalidCString",
+        Error::RollbackTransaction => "RollbackTransaction",
+        Error::AlreadyInTransaction => "AlreadyInTransaction",
+        Error::RollbackErrorOnCommit { .. } => "RollbackErrorOnCommit",
+        Error::NotInTransaction => "NotInTransaction",
+        Error::BrokenTransactionManager => "BrokenTransactionManager",
+        _ => "Other",
+    }
+}
+
+// This is a fake table for default fallback chunk size
+pub struct FakeTable {}
+impl field_count::FieldCount for FakeTable {
+    fn field_count() -> usize {
+        100
+    }
+}
+
+// A holder struct for processors db writing so we don't need to keep adding new params
+#[derive(Clone)]
+pub struct DbWriter {
```

Comment:
nit: can we use pub(crate) when possible?

Reply:
why?